// @agile-ts/core/dist/runtime/runtime.job.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeJob = void 0;
const internal_1 = require("../internal");
class RuntimeJob {
    constructor(observer, config = {}) {
        this.performed = false;
        this.subscriptionContainersToUpdate = new Set();
        this.triesToUpdate = 0;
        config = internal_1.defineConfig(config, {
            background: false,
            sideEffects: {
                enabled: true,
                exclude: [],
            },
            force: false,
            numberOfTriesToUpdate: 3,
        });
        this.config = {
            background: config.background,
            force: config.force,
            sideEffects: config.sideEffects,
            numberOfTriesToUpdate: config.numberOfTriesToUpdate,
        };
        this.observer = observer;
        this.rerender =
            !config.background &&
                this.observer.agileInstance().integrations.hasIntegration();
        this._key = config.key;
        this.subscriptionContainersToUpdate = new Set(observer.subs);
    }
    get key() {
        return this._key;
    }
    set key(value) {
        this._key = value;
    }
}
exports.RuntimeJob = RuntimeJob;
